### PR TITLE
v4l2decode: support wayland display

### PR DESCRIPTION
**Pull request by YuJiankang**
add wayland support to v4l2 decoder
Signed-off-by: jkyu <jiankang.yu@intel.com>

### CODE REVIEW

File: tests/v4l2decode.cpp [hunk 1]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
 
     return true;
 }
+
+#elif __ENABLE_WAYLAND__
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: 不同平台的接口类似，用类来实现可能更清晰一下。

File: tests/v4l2decode.cpp [hunk 2]
```diff
@@ -434,7 +581,7 @@ int main(int argc, char** argv)
     int32_t ioctlRet = -1;
     YamiMediaCodec::CalcFps calcFps;
 
-#if __ENABLE_X11__
+#if (defined (__ENABLE_X11__) && !defined(__ENABLE_WAYLAND__))
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: 将不同的平台作为互斥的选项可能会方便一些。 --enable-platform=x11/wayland/android

File: tests/v4l2decode.cpp [hunk 3]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+    output((struct wl_buffer *)(buf->m.userptr));
+    memset(&buffer, 0, sizeof(buffer));
+    buffer.type = V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE;
+    buffer.index = (buf->index) ? (buf->index - 1): (outputQueueCapacity - 1);
+    ioctlRet = SIMULATE_V4L2_OP(Ioctl)(fd, VIDIOC_QBUF, &buffer);
+    ASSERT(ioctlRet != -1);
+    return true;
+}
 #else
 bool displayOneVideoFrameEGL(int32_t fd, int32_t index)
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: 这个函数名称可以改一下了。

File: tests/v4l2decode.cpp (hunk 1 above)
Comment (xuguangxin) [inline tests/v4l2decode.cpp]: Agree, we need refact the code, not it's bit of mess here.

File: tests/v4l2decode.cpp [hunk 4]
```diff
@@ -36,6 +36,8 @@
 #if ANDROID
 #include <gui/SurfaceComposerClient.h>
 #include <va/va_android.h>
+#elif __ENABLE_WAYLAND__
+#include <va/va_wayland.h>
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: do we really use va_wayland.h here?

File: tests/v4l2decode.cpp [hunk 5]
```diff
@@ -537,7 +684,17 @@ int main(int argc, char** argv)
     }
     ioctlRet = SIMULATE_V4L2_OP(Ioctl)(fd, VIDIOC_S_FMT, &format);
     ASSERT(ioctlRet != -1);
-
+#if __ENABLE_WAYLAND__
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: how about move it above to be parallel with XOpenDisplay().

File: tests/v4l2decode.cpp [hunk 6]
```diff
@@ -715,7 +871,17 @@ int main(int argc, char** argv)
...
     }
 #endif
-
+#if __ENABLE_WAYLAND__
+    struct v4l2_buffer buffer;
+    //queue buffs
+    for (uint32_t i = 0; i < outputQueueCapacity; i++) {
+        memset(&buffer, 0, sizeof(buffer));
+        buffer.type = V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE;
+        buffer.index = i;
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: usually buffer.memory is required for VIDIOC_QBUF/VIDIOC_DQBUF.

File: tests/v4l2decode.cpp [hunk 7]
```diff
@@ -715,7 +871,17 @@ int main(int argc, char** argv)
         }
     }
 #endif
-
+#if __ENABLE_WAYLAND__
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: in fact, it does same as following "#ifndef ANDROID   takeOneOutputFrame(fd, i)"

File: tests/v4l2decode.cpp [hunk 8]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+    return vaPutSurfaceWayland(buffer, &srcRect, &dstRect);
+}
+
+static bool displayOneVideoFrameWayland(int32_t fd, struct v4l2_buffer  *buf)
+{
+    int32_t ioctlRet = -1;
+    struct v4l2_buffer buffer;
+
+    output((struct wl_buffer *)(buf->m.userptr));
+    memset(&buffer, 0, sizeof(buffer));
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: in fact, takeOneOutputFrame() does VIDIOC_QBUF.
we needn't QBUF here, nor for displayOneVideoFrameAndroid. (just be careful that buffer index may change).

File: tests/v4l2decode.cpp (hunk 6 above)
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: ok, buffer.memory is useless for wl_buffer type. we can ignore it.

Comment (halleyzhao) [inline tests/v4l2decode.cpp]: how about something like V4L2_MEMORY_DRIVERPTR? then we needn't make exception for the memory type elsewhere.

File: tests/v4l2decode.cpp [hunk 9]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+    buffer.type = V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE;
+    buffer.index = (buf->index) ? (buf->index - 1): (outputQueueCapacity - 1);
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: why minus one?

File: tests/v4l2decode.cpp [hunk 10]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+    SharedPtr<wl_compositor>     compositor;
+    SharedPtr<wl_shell>          shell;
+    SharedPtr<wl_shell_surface>  shell_surface;
+    SharedPtr<wl_surface>        surface;
+};
+
+static struct display waylandDisplay;
+static bool redrawPending = false;
+
+struct VADisplayTerminator {
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: is it really used? VADisplayTerminator

File: tests/v4l2decode.cpp [hunk 11]
```diff
@@ -200,15 +202,12 @@ static std::vector<uint8_t*> inputFrames;
 static std::vector<struct RawFrameData> rawOutputFrames;
 
 static FILE* outfp = NULL;
-#ifndef ANDROID
+
+#if (!defined(ANDROID) && !defined(__ENABLE_WAYLAND__))
```
Comment (xuguangxin) [inline tests/v4l2decode.cpp]: should be #if !__ENABLE_X11__

File: tests/v4l2decode.cpp [hunk 12]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+        if (!d->shell_surface)
+            return false;
+        wl_shell_surface_set_toplevel(d->shell_surface.get());
+    }
+    return true;
+}
+
+bool createWaylandDisplay()
+{
+    SharedPtr<VADisplay> display;
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: SharedPtr<VADisplay> display is not used.

File: tests/v4l2decode.cpp (hunk 1 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: yes, how about merge it firstly? and we take some time to refine the code in the future.

File: tests/v4l2decode.cpp (hunk 9 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: minus one, means the we can reuse the last surface for output.

File: tests/v4l2decode.cpp (hunk 5 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: ok, I will try move it to be parallel with XOpenDisplay().

File: tests/v4l2decode.cpp (hunk 4 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: yes, wayland related structure and interface are included by this header file.

File: tests/v4l2decode.cpp (hunk 12 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: yes, I will remove it, thanks

File: tests/v4l2decode.cpp (hunk 10 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: yes, I will remove it, thanks

File: tests/v4l2decode.cpp (hunk 4 above)
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: if wayland only, then wayland, not va/wayland.

File: tests/v4l2decode.cpp [hunk 13]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+    if (!ensureWindow(dstRect->width, dstRect->height))
+        return false;
+
+    wl_surface_attach(d->surface.get(), buffer, 0, 0);
+    wl_surface_damage(d->surface.get(), dstRect->x,
+		dstRect->y, dstRect->width, dstRect->height);
+    wl_display_flush(d->display.get());
+    redrawPending = true;
+    callback = wl_surface_frame(d->surface.get());
+    wl_callback_add_listener(callback, &frame_callback_listener,&redrawPending);
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: frame_callback_listener is local variable, i'm not sure whether it is ok to set it to a callback listener.

File: tests/v4l2decode.cpp [hunk 14]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
...
+    wl_surface_commit(d->surface.get());
+    return true;
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: seems, wl_buffer should be destroyed on a callback. otherwise, some mem leak.

File: tests/v4l2decode.cpp (hunk 4 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: ok, I will try include wayland-client.h, thanks

File: tests/v4l2decode.cpp (hunk 14 above)
Comment (YuJiankang) [inline tests/v4l2decode.cpp]: yes, may be a problem, I will try to destroy the wl_buffer

File: tests/v4l2decode.cpp [hunk 15]
```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
 
     return true;
 }
+
+#elif __ENABLE_WAYLAND__
+struct display {
+    SharedPtr<wl_display>        display;
```
Comment (halleyzhao) [inline tests/v4l2decode.cpp]: not so good, you can't decide the order to destroy these object.
if wl_display is destroyed before wl_surface, there will be exception.

**Comment by halleyzhao** [comment]
it's ok to merge first, then does some refine work later.

**Comment by xuguangxin** [comment]
+1, thanks Halley for review. 
merged as 896be8bd26c119d36da0fe58cafb280a1499806b

**Pull request closed without merging**